### PR TITLE
Fix Client#createGuild() bug where shard was undefined

### PR DESCRIPTION
**Pull request by apacheli**
This fixes 2 issues that were recently discovered in the Eris channel.

1. The first issue was that it was a race condition between `Guild#shard` and `Client#guildCreate`.
2. The second issue comes from the first one. The guild obtained from REST was not cached.

### CODE REVIEW

**Comment by Skillz4Killz** [comment]
Why not have a property dedicated to the client? `this.client = client;`. You'd never have bugs like this again and users don't need to go `guild.shard.client`. One way or the other, `client` exists in the class.

Note: This needs typings updated to reflect the changes in the PR.

File: lib/structures/Guild.js [hunk 1]
```diff
@@ -50,7 +50,10 @@ const VoiceState = require("./VoiceState");
 class Guild extends Base {
     constructor(data, client) {
         super(data.id);
-        this.shard = client.shards.get(client.guildShardMap[this.id]);
+        this._client = client;
+        this.shard = client.shards.get(client.guildShardMap[this.id]) || {
```
Comment (abalabahaha) [inline lib/structures/Guild.js]: `||` is kind of unnecessary with `_client`

File: lib/structures/GuildChannel.js [hunk 2]
```diff
@@ -23,6 +23,9 @@ class GuildChannel extends Channel {
     constructor(data, client) {
         super(data);
         this.guild = client.guilds.get(data.guild_id) || {
+            shard: {
```
Comment (abalabahaha) [inline lib/structures/GuildChannel.js]: Could use `guild._client` here and this wouldn't be necessary either.

**Review by abalabahaha — CHANGES_REQUESTED** [review]

**Comment by apacheli** [comment]
Sorry for delaying so long. I completely forgot about this. Happy late holidays lol

**Review by abalabahaha — APPROVED** [review]

**Pull request merged**